### PR TITLE
Rtt print

### PR DESCRIPTION
**Pull request by Michel-Haber**
There is no issue for this.
This PR gives a fully functioning debug macro AMPEROSE_DEBUG.
It works using the SEGGER RTT (Check licensing for this...)
The macro function works exactly like printf, and can be deactivated by setting DEBUG to 0.

### CODE REVIEW

File: microcontroller/include/rtt.h [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#ifndef RTT_H
+#define RTT_H
+
+#include <stdint.h>
+#include <stdlib.h>
+
+#include "hal.h"
```
Comment (abdelhadi88) [inline microcontroller/include/rtt.h]: Why is this included ?

File: microcontroller/src/rtt.c [hunk 2]
```diff
@@ -0,0 +1,42 @@
+#include "rtt.h"
+
+// This stream is to be used for all communication with RTT
+BaseSequentialStream RTT_stream;
```
Comment (abdelhadi88) [inline microcontroller/src/rtt.c]: Again, why are you defining a Base Sequential Stream ?

File: microcontroller/src/rtt.c [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+static size_t RTT_write(void *instance, const uint8_t *bp, size_t n){
+  (void) instance;
+  return SEGGER_RTT_Write(0, (void*)bp, (unsigned)n);
+}
+
+// Note that the read wait here is a sleep in order not to hinder threads of lower priority
+static size_t RTT_read(void *instance, uint8_t *bp, size_t n){
+  (void) instance;
+  while(!SEGGER_RTT_HasKey()){
+    chThdSleepMilliseconds(50);
```
Comment (abdelhadi88) [inline microcontroller/src/rtt.c]: Seriously ?

**Review by abdelhadi88 — COMMENTED** [review]
Although you removed them in the next commit, anything related to GPL functions must be removed entirely from the history. Please squash both commits into one

File: microcontroller/include/debug.h [hunk 4]
```diff
@@ -0,0 +1,10 @@
+#ifndef DEBUG_H
+#define DEBUG_H
+
+#include "SEGGER_RTT.h"
+
+#define DEBUG 1
```
Comment (abdelhadi88) [inline microcontroller/include/debug.h]: Wouldn't it be better to add this line to an amperose_conf.h file where all configurations are regrouped ?

**Comment by abdelhadi88** [comment]
Another comment: This PR does have an issue (#3).

File: microcontroller/src/rtt.c [hunk 5]
```diff
@@ -0,0 +1,18 @@
+#include "rtt.h"
+
+size_t RTT_write(const uint8_t *bp, size_t n){
+  return SEGGER_RTT_Write(0, (void*)bp, (unsigned)n);
+}
+
+// Note that the read wait here is a sleep in order not to hinder threads of lower priority
+size_t RTT_read(uint8_t *bp, size_t n){
+  while(!SEGGER_RTT_HasKey()){
+  //  chThdSleepMilliseconds(50);
```
Comment (abdelhadi88) [inline microcontroller/src/rtt.c]: Please remove this line

File: microcontroller/src/rtt.c [hunk 6]
```diff
@@ -0,0 +1,18 @@
+#include "rtt.h"
+
+size_t RTT_write(const uint8_t *bp, size_t n){
```
Comment (abdelhadi88) [inline microcontroller/src/rtt.c]: It is up to you but it would be good if you insert a space between ') {' (because this is the style I am using)

File: microcontroller/src/rtt.c [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+  return SEGGER_RTT_Write(0, (void*)bp, (unsigned)n);
+}
+
+// Note that the read wait here is a sleep in order not to hinder threads of lower priority
+size_t RTT_read(uint8_t *bp, size_t n){
```
Comment (abdelhadi88) [inline microcontroller/src/rtt.c]: Same comment about coding style ') {'

File: microcontroller/src/rtt.c [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+  while(!SEGGER_RTT_HasKey()){
+  //  chThdSleepMilliseconds(50);
+  }
+  return SEGGER_RTT_Read(0, (void*)bp, (unsigned)n);
+}
+
+// RTT initialization
+void RTT_init(void){
```
Comment (abdelhadi88) [inline microcontroller/src/rtt.c]: Same comment about coding style ') {'

File: microcontroller/include/debug.h (hunk 4 above)
Comment (Michel-Haber) [reply microcontroller/include/debug.h]: I don't mind, I'll move it

File: microcontroller/src/rtt.c (hunk 5 above)
Comment (Michel-Haber) [reply microcontroller/src/rtt.c]: Ok gonna amend

**Review by abdelhadi88 — APPROVED** [review]
